property.go

```diff
@@ -87,10 +88,132 @@ func trimUT8StringUpTo(maxLength int, s string) string {
 	return s[:length]
 }
 
+func (p *BaseProperty) GetValueType() ValueDataType {
+	for k, v := range p.ICalParameters {
+		if Parameter(k) == ParameterValue && len(v) == 1 {
+			return ValueDataType(v[0])
+		}
+	}
+
+	// defaults from spec if unspecified
+	switch Property(p.IANAToken) {
+	case PropertyCalscale:
+		fallthrough
+	case PropertyMethod:
+		fallthrough
+	case PropertyProductId:
+		fallthrough
+	case PropertyVersion:
+		fallthrough
+	case PropertyCategories:
+		fallthrough
+	case PropertyClass:
+		fallthrough
+	case PropertyComment:
+		fallthrough
+	case PropertyDescription:
+		fallthrough
+	case PropertyLocation:
+		fallthrough
+	case PropertyResources:
+		fallthrough
+	case PropertyStatus:
+		fallthrough
+	case PropertySummary:
+		fallthrough
+	case PropertyTransp:
+		fallthrough
+	case PropertyTzid:
+		fallthrough
+	case PropertyTzname:
+		fallthrough
+	case PropertyContact:
+		fallthrough
+	case PropertyRelatedTo:
+		fallthrough
+	case PropertyUid:
+		fallthrough
+	case PropertyAction:
+		fallthrough
+	default:
+		fallthrough
+	case PropertyRequestStatus:
+		return ValueDataTypeText
+
+	case PropertyAttach:
+		fallthrough
+	case PropertyTzurl:
+		fallthrough
+	case PropertyUrl:
+		return ValueDataTypeUri
+
+	case PropertyGeo:
+		return ValueDataTypeFloat
+
+	case PropertyPercentComplete:
+		fallthrough
+	case PropertyPriority:
+		fallthrough
+	case PropertyRepeat:
+		fallthrough
+	case PropertySequence:
+		return ValueDataTypeInteger
+
+	case PropertyCompleted:
+		fallthrough
+	case PropertyDtend:
+		fallthrough
+	case PropertyDue:
+		fallthrough
+	case PropertyDtstart:
+		fallthrough
+	case PropertyRecurrenceId:
+		fallthrough
+	case PropertyExdate:
+		fallthrough
+	case PropertyRdate:
+		fallthrough
+	case PropertyCreated:
+		fallthrough
+	case PropertyDtstamp:
+		fallthrough
+	case PropertyLastModified:
+		return ValueDataTypeDateTime
+
+	case PropertyDuration:
+		fallthrough
+	case PropertyTrigger:
+		return ValueDataTypeDuration
+
+	case PropertyFreebusy:
+		return ValueDataTypePeriod
+
+	case PropertyTzoffsetfrom:
+		fallthrough
+	case PropertyTzoffsetto:
+		return ValueDataTypeUtcOffset
+
+	case PropertyAttendee:
+		fallthrough
+	case PropertyOrganizer:
+		return ValueDataTypeCalAddress
+
+	case PropertyRrule:
+		return ValueDataTypeRecur
+	}
+}
+
 func (property *BaseProperty) serialize(w io.Writer) {
 	b := bytes.NewBufferString("")
 	fmt.Fprint(b, property.IANAToken)
-	for k, vs := range property.ICalParameters {
+
+	var keys []string
+	for k := range property.ICalParameters {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
```

Comment:
Is the basis of this sort 

https://datatracker.ietf.org/doc/html/rfc5545#section-3.8.2.6

I believe my intention here was to maintain the input order. However nothing stops us inserting the params in the first best match or some such.

Reply:
Yeah I mentioned in the PR description but the sort is to make the serialization deterministic. Mainly for testing. 

Because the property params are saved in a map in the deserialized model the original order is not currently preserved, and we get a random output order each time. 

I suppose also saving the original order of the keys is a possibility, seems cleaner.

Reply:
Yeah I would accept that. But you make a good point about it being in a map. Since you're doing the grunt work here I will let you propose what we do about this, but not determinism is a massive oversight of mine. Sorry about that.

Haha. Yep. From day 1:

![image](https://github.com/arran4/golang-ical/assets/111667/614afc3c-676c-413f-9268-f336b5cffe7b)

But in short, I no longer have an issue with this. I guess respond to the others, simplify the switch and we can go from there.

Reply:
I started looking into this one, I wouldn't want to introduce a breaking change, so if we did want to preserve original order, I'd imagine we'd introduce a new field:
```
ICalParameterKeys []string // used to preserve original key order
ICalParameters        map[string][]string
```

Reply:
Yeah I don't think there is a way we can preserve order without a major change here. I think between:
1. Random ordering as it was
2. Deterministic ordering 
3. Breaking api

We should just stick to your change here. Number 2.

Sometimes I wish you could turn features on and off on maps, such as "preserve order"